// hardhat/dist/src/internal/builtin-plugins/gas-analytics/hook-handlers/test.js
import { getGasAnalyticsManager } from "../helpers.js";
export default async () => ({
    onTestRunStart: async (context, id, next) => {
        await next(context, id);
        await testRunStart(context, id);
    },
    onTestWorkerDone: async (context, id, next) => {
        await next(context, id);
        await testWorkerDone(context, id);
    },
    onTestRunDone: async (context, id, next) => {
        await next(context, id);
        await testRunDone(context, id);
    },
});
function isGasStatsEnabled(context) {
    return (context.globalOptions.gasStats === true ||
        context.globalOptions.gasStatsJson !== undefined);
}
export async function testRunStart(context, id) {
    if (isGasStatsEnabled(context)) {
        await getGasAnalyticsManager(context).clearGasMeasurements(id);
    }
}
export async function testWorkerDone(context, id) {
    if (isGasStatsEnabled(context)) {
        await getGasAnalyticsManager(context).saveGasMeasurements(id);
    }
}
export async function testRunDone(context, id) {
    if (context.globalOptions.gasStats === true) {
        await getGasAnalyticsManager(context).reportGasStats(id);
    }
    if (context.globalOptions.gasStatsJson !== undefined) {
        await getGasAnalyticsManager(context).writeGasStatsJson(context.globalOptions.gasStatsJson, id);
    }
}
