// hardhat/dist/src/internal/builtin-plugins/network-manager/edr/types/hardfork.js
import { l1HardforkLatest, opLatestHardfork } from "@nomicfoundation/edr";
import { OPTIMISM_CHAIN_TYPE } from "../../../../constants.js";
import { edrL1HardforkToHardhatL1HardforkName, edrOpHardforkToHardhatOpHardforkName, } from "../utils/convert-to-edr.js";
export var L1HardforkName;
(function (L1HardforkName) {
    L1HardforkName["FRONTIER"] = "chainstart";
    L1HardforkName["HOMESTEAD"] = "homestead";
    L1HardforkName["DAO"] = "dao";
    L1HardforkName["TANGERINE_WHISTLE"] = "tangerineWhistle";
    L1HardforkName["SPURIOUS_DRAGON"] = "spuriousDragon";
    L1HardforkName["BYZANTIUM"] = "byzantium";
    L1HardforkName["CONSTANTINOPLE"] = "constantinople";
    L1HardforkName["PETERSBURG"] = "petersburg";
    L1HardforkName["ISTANBUL"] = "istanbul";
    L1HardforkName["MUIR_GLACIER"] = "muirGlacier";
    L1HardforkName["BERLIN"] = "berlin";
    L1HardforkName["LONDON"] = "london";
    L1HardforkName["ARROW_GLACIER"] = "arrowGlacier";
    L1HardforkName["GRAY_GLACIER"] = "grayGlacier";
    L1HardforkName["MERGE"] = "merge";
    L1HardforkName["SHANGHAI"] = "shanghai";
    L1HardforkName["CANCUN"] = "cancun";
    L1HardforkName["PRAGUE"] = "prague";
    L1HardforkName["OSAKA"] = "osaka";
})(L1HardforkName || (L1HardforkName = {}));
export var OpHardforkName;
(function (OpHardforkName) {
    OpHardforkName["BEDROCK"] = "bedrock";
    OpHardforkName["REGOLITH"] = "regolith";
    OpHardforkName["CANYON"] = "canyon";
    OpHardforkName["ECOTONE"] = "ecotone";
    OpHardforkName["FJORD"] = "fjord";
    OpHardforkName["GRANITE"] = "granite";
    OpHardforkName["HOLOCENE"] = "holocene";
    OpHardforkName["ISTHMUS"] = "isthmus";
})(OpHardforkName || (OpHardforkName = {}));
const L1_HARDFORK_ORDER = Object.values(L1HardforkName);
const OP_HARDFORK_ORDER = Object.values(OpHardforkName);
export function getHardforks(chainType) {
    return chainType === OPTIMISM_CHAIN_TYPE
        ? OP_HARDFORK_ORDER
        : L1_HARDFORK_ORDER;
}
export function getCurrentHardfork(chainType) {
    return chainType === OPTIMISM_CHAIN_TYPE
        ? edrOpHardforkToHardhatOpHardforkName(opLatestHardfork())
        : edrL1HardforkToHardhatL1HardforkName(l1HardforkLatest());
}
/**
 * Check if `hardforkA` is greater than or equal to `hardforkB`,
 * that is, if it includes all its changes.
 *
 * This function is not type-safe, as it accepts any string as hardfork name.
 * It is the caller's responsibility to ensure that the hardfork names are valid.
 */
export function hardforkGte(hardforkA, hardforkB, chainType) {
    return chainType === OPTIMISM_CHAIN_TYPE
        ? opHardforkGte(hardforkA, hardforkB)
        : l1HardforkGte(hardforkA, hardforkB);
}
function opHardforkGte(hardforkA, hardforkB) {
    /* eslint-disable-next-line @typescript-eslint/consistent-type-assertions
    -- Cast is safe, as the function is only called with valid hardfork names. */
    const indexA = OP_HARDFORK_ORDER.indexOf(hardforkA);
    /* eslint-disable-next-line @typescript-eslint/consistent-type-assertions
    -- Cast is safe, as the function is only called with valid hardfork names. */
    const indexB = OP_HARDFORK_ORDER.indexOf(hardforkB);
    return indexA >= indexB;
}
function l1HardforkGte(hardforkA, hardforkB) {
    /* eslint-disable-next-line @typescript-eslint/consistent-type-assertions
    -- Cast is safe, as the function is only called with valid hardfork names. */
    const indexA = L1_HARDFORK_ORDER.indexOf(hardforkA);
    /* eslint-disable-next-line @typescript-eslint/consistent-type-assertions
    -- Cast is safe, as the function is only called with valid hardfork names. */
    const indexB = L1_HARDFORK_ORDER.indexOf(hardforkB);
    return indexA >= indexB;
}
export function isValidHardforkName(hardfork, chainType) {
    return chainType === OPTIMISM_CHAIN_TYPE
        ? /* eslint-disable-next-line @typescript-eslint/consistent-type-assertions
          -- Cast is safe, as we're validating the hardfork name. */
            OP_HARDFORK_ORDER.includes(hardfork)
        : /* eslint-disable-next-line @typescript-eslint/consistent-type-assertions
          -- Cast is safe, as we're validating the hardfork name. */
            L1_HARDFORK_ORDER.includes(hardfork);
}
