// hardhat/dist/src/types/solidity/build-system.js
export var CompilationJobCreationErrorReason;
(function (CompilationJobCreationErrorReason) {
    /**
     * The root file's own pragmas are incompatible with all configured compilers.
     */
    CompilationJobCreationErrorReason["NO_COMPATIBLE_SOLC_VERSION_WITH_ROOT"] = "NO_COMPATIBLE_SOLC_VERSION_WITH_ROOT";
    /**
     * A dependency's own pragmas are incompatible with all configured compilers.
     */
    CompilationJobCreationErrorReason["NO_COMPATIBLE_SOLC_VERSION_WITH_DEPENDENCY"] = "NO_COMPATIBLE_SOLC_VERSION_WITH_DEPENDENCY";
    /**
     * Root and a transitive import path have contradictory pragmas (invalid range / empty intersection).
     */
    CompilationJobCreationErrorReason["IMPORT_OF_INCOMPATIBLE_FILE"] = "IMPORT_OF_INCOMPATIBLE_FILE";
    /**
     * Root and a transitive import path have a valid range but no configured compiler satisfies it.
     */
    CompilationJobCreationErrorReason["NO_COMPATIBLE_SOLC_VERSION_FOR_TRANSITIVE_IMPORT_PATH"] = "NO_COMPATIBLE_SOLC_VERSION_FOR_TRANSITIVE_IMPORT_PATH";
    /**
     * The override version doesn't satisfy the root file's own pragmas.
     */
    CompilationJobCreationErrorReason["INCOMPATIBLE_OVERRIDDEN_SOLC_VERSION"] = "INCOMPATIBLE_OVERRIDDEN_SOLC_VERSION";
    /**
     * A dependency's pragmas are incompatible with the override version.
     */
    CompilationJobCreationErrorReason["OVERRIDDEN_SOLC_VERSION_INCOMPATIBLE_WITH_DEPENDENCY"] = "OVERRIDDEN_SOLC_VERSION_INCOMPATIBLE_WITH_DEPENDENCY";
    /**
     * Generic fallback — no single compiler works for root + all dependencies.
     */
    CompilationJobCreationErrorReason["NO_COMPATIBLE_SOLC_VERSION_FOUND"] = "NO_COMPATIBLE_SOLC_VERSION_FOUND";
})(CompilationJobCreationErrorReason || (CompilationJobCreationErrorReason = {}));
/**
 * The restult of building a file.
 */
export var FileBuildResultType;
(function (FileBuildResultType) {
    FileBuildResultType["CACHE_HIT"] = "CACHE_HIT";
    FileBuildResultType["BUILD_SUCCESS"] = "BUILD_SUCCESS";
    FileBuildResultType["BUILD_FAILURE"] = "BUILD_FAILURE";
})(FileBuildResultType || (FileBuildResultType = {}));
